Make determine_version compatible with python 2 or 3

nickanderson opened the pull request:
Changelog: Title

nickanderson commented:
@BasementTrix sent me this patch so I made a PR for it.

nickanderson commented:
@BasementTrix you want the history to reflect "Trix Farrar"?

BasementTrix commented:
On Mon, Mar 25, 2019 at 07:54:07AM -0700, Nick Anderson wrote:
> @BasementTrix you want the history to reflect "Trix Farrar"?
> 

Please.

-- 
John D. "Trix" Farrar            __\\|//__               Basement.NET
trix@basement.net                (` o-o ')   http://www.basement.net/
--------------------------------ooO-(_)-Ooo--------------------------
 GPG Key Fprint: 39D0 BCD2 9C9F B7EF B125  D419 61E8 8648 16F2 1814

nickanderson commented:
You want `John D. "Trix" Farrar` or just Trix Farrar?

olehermanse commented:
This pull request **introduces 2 alerts** when merging 431b421583af9f3d0cbd15f473a9954a1ed3757c into 3fb84b7d35b86e75b20d8616ca275cb8130714f6 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-f47c1bfea1a7f2488a53a083300dffbb1041ba59)

**new alerts:**

* 1 for Unused import
* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

olehermanse commented on misc/determine-version.py:
```diff
@@ -1,11 +1,13 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2.7
```
This change seems like it could break it for others.

BasementTrix replied:
On the Linux boxen I've checked, the python package installed python2.7 as the actual executable and python2 and python as links to it.  I could be wrong.

BasementTrix replied:
The RPMs that I've reviewed install python2.7 as the actual file with python2 and python as symlinks to it.

olehermanse replied:
I think leaving it as 

```
#!/usr/bin/env python
```

And making sure we are compatible with both 2.7 and 3+ is the safest way.

vpodzime replied:
> And making sure we are compatible with both 2.7 and 3+ is the safest way.

:+1:

olehermanse commented on misc/determine-version.py:
```diff
@@ -1,11 +1,13 @@
 import os
 from os.path import basename
+import  functools
+from functools import cmp_to_key
```
Don't import functools twice. Omit line 9.

BasementTrix replied:
You are correct.  Still a python neophyte.

BasementTrix commented:
> You want `John D. "Trix" Farrar` or just Trix Farrar?

I guess `John D. "Trix" Farrar`, since that's how I'm credited in the pkg module.

BasementTrix commented on .mailmap:
```diff
@@ -19,3 +19,4 @@ Chris Dituri <csdituri@gmail.com> chris dituri
 Ole Herman S. Elgesem <ole@northern.tech> <ole.elgesem@northern.tech> <ole.elgesem@cfengine.com>
 Vratislav Podzimek <vratislav.podzimek@northern.tech> <vratislav.podzimek@cfengine.com>
 Craig Comstock <craig.comstock@northern.tech> <craig.comstock@cfengine.com>
+Trix Farrar <trix@basement.net> <trix@doyle.basement.net>
```
trix@doyle.basement.net address is not valid.  s/b

Trix Farrar <trix@basement.net> <trixfarrar@gmail.com>

nickanderson replied:
> > You want `John D. "Trix" Farrar` or just Trix Farrar?
> I guess `John D. "Trix" Farrar`, since that's how I'm credited in the pkg module.

Well, we can align them in the .mailmap if you prefer one over another.

vpodzime commented on misc/determine-version.py:
```diff
@@ -139,10 +141,10 @@ def version_cmp(a, b):
 git = subprocess.Popen(["git", "rev-list", "-1", REV],
                        stdout=subprocess.PIPE)
-full_rev = git.stdout.readlines()[0].strip()
+full_rev = git.stdout.readlines()[0].strip().decode()
```
I think `decode()` should be done before `strip()`. Hopefully, *UTF-8* is designed in a way that there can be no difference when changing the order here, but the golden rule is to convert the data ASAP and further process it.

nickanderson replied:
I just added a separate commit for this, pushed.

nickanderson commented:
@BasementTrix sent me a new patch against master to address some comments. I just re-wrote the history on this branch so sorry if it complicates the review.

- I changed the author to `John D. "Trix"  Farrar <trix@basement.net>'
- I added trixfarrar@gmail.com as an alias to the .mailmap

olehermanse commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 2ec1bd36e6c430cbf7b4d083144309dc540aa94c into 73167a7efd4edb90b456251bb8429eddf0356ee5 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-8e01fc5041145afd763f65d5adc4699dfb19d3eb)

**new alerts:**

* 1 for Syntax error

**fixed alerts:**

* 1 for File is not always closed

---

*Comment posted by [LGTM.com](https://lgtm.com)*

BasementTrix commented:
> * 1 for Syntax error

I'm testing under Python 3.6 and have not seen this error.

[trix@doyle core]$ python misc/determine-version.py                                                                   
determine-version.py: REV        = HEAD
determine-version.py: full_rev   = f9652c5fdefc0aca9255d996763b91129c5bcaab
determine-version.py: abbrev_rev = f9652c5fd
determine-version.py: exact_tag  = not found
determine-version.py: recent_tag = 3.12.0b1
determine-version.py: recent_rev = 2476fc575060663b5a5645971ae531ae899ca77b
determine-version.py: recent_version =  ('3', '12', '0', 'b1')
determine-version.py: all_tags   = [('3', '12', '0', ''), ('3', '12', '0', 'b1-build2'), ('3', '12', '0', 'b1'), ('3', '12', '0', '-build6'), ('3', '12', '0', '-build5'), ('3', '12', '0', '-build4'), ('3', '12', '0', '-build3'), ('3', '12', '0', '-build2'), ('3', '12', '0', '-build1'), ('3', '12', '0', '-3756')]
determine-version.py: SUCCESS, version detected is: 3.13.0a.f9652c5fd
3.13.0a.f9652c5fd
[trix@doyle core]$

nickanderson commented:
@cf-bottom jenkins, please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2227)](https://ci.cfengine.com//job/pr-pipeline/2227/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2227/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2227/

BasementTrix commented:
The same change will need to be made in the 3rdparty/core/determine-version.py file in cfengine/masterfiles because it is a copy of misc/determine-version.py from here (cfengine/core), if/when this change is merged.

olehermanse commented on misc/determine-version.py:
```diff
@@ -139,10 +140,10 @@ def version_cmp(a, b):
 git = subprocess.Popen(["git", "rev-list", "-1", "--abbrev-commit", REV],
                        stdout=subprocess.PIPE)
-abbrev_rev = git.stdout.readlines()[0].strip()
+abbrev_rev = git.stdout.readlines()[0].strip().decode()
```
It seems more correct to decode and then strip here as well, and all places below. (?)

nickanderson replied:
OK, globally replaced.

BasementTrix replied:
Blame it on that Trix guy.  He doesn't know any better.

olehermanse commented:
@nickanderson please squash

nickanderson commented:
@olehermanse as you wish

olehermanse commented:
@cf-bottom Let's see if jenkins likes this again.

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2300)](https://ci.cfengine.com//job/pr-pipeline/2300/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2300/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2300/
